[WIP] Add Alpine linux to the GitLab CI pipeline

D4N opened the pull request:
This required some refactoring of the scripts in `contrib/vms/` as synced folders with vagrant are broken on alpine linux. There is now a little duplication in the scripts, but I think it is minimal and still bearable.

Unfortunately, the tests are broken on alpine, as they use busybox' diff, which has fewer features then gnu diff. Our testsuite relies on the option `ignore-matching-lines=^` to exist, which doesn't in busybox.

D4N committed:
[CI] Bump Fedora version on GitLab & in the Vagrant VMs

D4N committed:
[CI] Remove utils.source in the Vagrant provisioning scripts

=> we no longer need the shared volume, which is broken on Alpine Linux

D4N committed:
[CI] Add alpine Linux to Gitlab CI & Vagrant VMs

- need to disable synced folders in vagrant, as that feature is broken on Alpine
- need to build pip3 manually, as it is (apparently) not present in the Alpine
  repos

Additional packages for alpine:
- gtest-dev does not automatically pull in gtest (which we need)
- bash is not present in the docker image by default
- libxml2-utils for xmllint
- diffutils for GNU diff (required for the bash based test suite)

piponazo reviewed (dismissed):
Great! 

I was actually trying to do the same yesterday with Vagrant :grin: .

piponazo commented:
There is a failure on the new image:

```
/bin/sh: eval: line 70: ci/install_dependencies.sh: not found
Uploading artifacts...
```
Weird ... I have approved taking into account that you will merge only once the CI checks are green.

D4N commented:
Yeah, I think the alpine image has been seriously stripped down, so that even a shell is missing (then again it's only a few megabytes!). I'll try to fix that and won't merge this before that.

clanmills commented:
Alpine offers several shell/environments.  I selected "busy-box" because I've used it a lot on embedded controllers.  It has a "OK" shell, however it's not bash.  It's an OK shell.  My understanding is that the shell/command-interpreter has most of the linux utilities built in.  Bash has builtin's such as cd, echo and so on.  Builtins are fast because you don't need to fork a process.  So it has a smaller run-time foot-print.  And it start new command very quickly.

I didn't have any difficulty getting Alpine installed and building Exiv2.  However you may need a special script to install the dependencies on Alpine.

If this is troublesome (or you're too busy to work on this), I'll get Jenkins to build on Alpine via ssh.  I expect that to be straightforward because the Alpine VM is already configured.

D4N commented:
> Alpine offers several shell/environments.

Yeah, unfortunately one seems to need one for gitlab CI to run, but you can't use gitlab CI to select one, if it can't run. I guess there's a simple fix, we can't be the only people using the Alpine image from Dockerhub on Gitlab.

D4N committed:
[CI] change ci/install_dependencies shebang from bash to sh

- the script needs to be able to bootstrap everything even on platforms where we
  don't have bash
- drop function which is bash specific

D4N committed:
[unittests] drop linking against GTest::Main

- we don't need to link against it, as we provide our own main
- this breaks platforms that don't ship gtest_main.so (e.g. Alpine)

D4N committed:
[unittests] Fix strerror test in test_futils.cpp for musl libc

D4N commented:
The tests are now passing, please don't merge this yet, I want to rebase the commits first.

D4N commented:
@piponazo Could you take a look again, I had to do some tiny code changes.

piponazo commented on unitTests/CMakeLists.txt:
```diff
@@ -22,7 +22,6 @@ target_link_libraries(unit_tests
     PRIVATE
         exiv2lib
         GTest::GTest
-        GTest::Main
```
Good to know. I used GTest for years, and I never took the time to check whether gtest_main was needed or not.

D4N replied:
You can use `GTest::Main`, then you don't have to provide a `main()` that runs all tests yourself. But that requires that `gtest_main.a/.so` exists, which it doesn't on Alpine, so I'd prefer to keep it the way it is.

piponazo replied:
Sure. What I meant is that I always include it without thinking what it was. I actually always provide my own main function, specially in big projects where you need to do advance stuff ;)

piponazo commented:
Everything is looking fine for me ;)
